File: evals/python/report.py
from dataclasses import dataclass


@dataclass(frozen=True)
class Interval:
    """Half-open [start, end) in whole minutes since midnight."""
    start: int
    end: int


def render_row(label, interval, unit, precision, align, width, fill):
    span = interval.end - interval.start
    text = f"{span:.{precision}f}{unit}"
    if align == "right":
        return (label + text).rjust(width, fill)
    return (label + text).ljust(width, fill)


def classify(interval, now, holidays):
    if interval.end <= interval.start:
        raise ValueError(f"interval ends before it starts: {interval}")
    if interval.start > now:
        return "future"
    if interval.start in holidays:
        return "holiday"
    return "past"

File: evals/ts/report.ts
import { Order } from "./orders";

const MAX_RETRIES = 3;

export function renderRow(
  label: string,
  order: Order,
  unit: string,
  precision: number,
  align: string,
  width: number,
  fill: string,
): string {
  const text = `${order.amount.toFixed(precision)}${unit}`;
  if (align === "right") {
    return (label + text).padStart(width, fill);
  }
  return (label + text).padEnd(width, fill);
}

export function totals(orders: Order[]): Record<string, number> {
  const byId: Record<string, number> = {};
  for (const order of orders) {
    byId[order.id] = (byId[order.id] ?? 0) + order.amount;
  }
  return byId;
}

export function attempts(): number {
  return MAX_RETRIES;
}

export function summarise(orders: Order[]): string {
  const ids = new Set(orders.map((o) => o.id));
  return [...ids].join(",");
}

export function load(raw: string): Order[] {
  try {
    return JSON.parse(raw) as Order[];
  } catch (e) {}
  return [];
}
